cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/expressions/ComprehensionExpression.kt

```diff
@@ -0,0 +1,126 @@
+ *                   $$ |      $$$$$$$  |$$ |$$$$\
+ *                   $$ |      $$  ____/ $$ |\_$$ |
+ *                   $$ |  $$\ $$ |      $$ |  $$ |
+ *                   \$$$$$   |$$ |      \$$$$$   |
+ *                    \______/ \__|       \______/
+ *
+ */
+package de.fraunhofer.aisec.cpg.graph.statements.expressions
+
+import de.fraunhofer.aisec.cpg.graph.AccessValues
+import de.fraunhofer.aisec.cpg.graph.ArgumentHolder
+import de.fraunhofer.aisec.cpg.graph.edges.ast.astEdgeOf
+import de.fraunhofer.aisec.cpg.graph.edges.ast.astOptionalEdgeOf
+import de.fraunhofer.aisec.cpg.graph.edges.unwrapping
+import de.fraunhofer.aisec.cpg.graph.statements.Statement
+import java.util.Objects
+import kotlin.collections.plusAssign
+import org.apache.commons.lang3.builder.ToStringBuilder
+import org.neo4j.ogm.annotation.Relationship
+
+/** This class holds the variable, iterable and predicate of the [CollectionComprehension]. */
+class ComprehensionExpression : Expression(), ArgumentHolder {
+    @Relationship("VARIABLE")
+    var variableEdge =
+        astEdgeOf<Statement>(
+            of = ProblemExpression("Missing variableEdge in ${this::class}"),
+            onChanged = { _, new ->
+                val end = new?.end
+                if (end is Reference) {
+                    end.access = AccessValues.WRITE
+                }
+            }
+        )
+
+    /**
+     * This field contains the iteration variable of the comprehension. It can be either a new
+     * variable declaration or a reference (probably to a new variable).
+     */
+    var variable by unwrapping(ComprehensionExpression::variableEdge)
+
+    @Relationship("ITERABLE")
+    var iterableEdge =
+        astEdgeOf<Expression>(ProblemExpression("Missing iterable in ${this::class}"))
+
+    /** This field contains the iteration subject of the loop. */
+    var iterable by unwrapping(ComprehensionExpression::iterableEdge)
+
+    @Relationship("PREDICATE") var predicateEdge = astOptionalEdgeOf<Statement>()
+
+    /**
+     * This field contains the predicate which has to hold to evaluate `statement(variable)` and
+     * include it in the result.
+     */
+    var predicate by unwrapping(ComprehensionExpression::predicateEdge)
+
+    override fun toString() =
+        ToStringBuilder(this, TO_STRING_STYLE)
+            .appendSuper(super.toString())
+            .append("variable", variable)
+            .append("iterable", iterable)
+            .append("predicate", predicate)
+            .toString()
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is ComprehensionExpression) return false
+        return super.equals(other) &&
+            variable == other.variable &&
+            iterable == other.iterable &&
+            predicate == other.predicate
+    }
+
+    override fun hashCode() = Objects.hash(super.hashCode(), variable, iterable, predicate)
+
+    override fun addArgument(expression: Expression) {
```

Comment:
I don't get the need for this function.

Reply:
Required by implementing ArgumentHolder. Adds stuff for Fluent DSL.